Editorial review: Add reading-flow and reading-order docs (#39404)
mdn/content: files/en-us/web/css/reading-order/index.md

---
title: reading-order
slug: Web/CSS/reading-order
page-type: css-property
browser-compat: css.properties.reading-order
---

{{CSSRef}}

> The **`reading-order`** [CSS](/en-US/docs/Web/CSS) property sets a {{glossary("reading order")}} ordinal group value on elements within a [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) container, enabling authors to override the [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) set via the container's [reading-flow](/en-US/docs/Web/CSS/reading-flow) property value.

Reviewer: ```suggestion
The **`reading-order`** [CSS](/en-US/docs/Web/CSS) property enables defining a {{glossary("reading order")}} for elements within a [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) container, which overrides any [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) set on the container by its [reading-flow](/en-US/docs/Web/CSS/reading-flow) property.
```
The first sentence is what it does, not how to do it. Let's add the details in the description.

Author: updated

## Syntax

```css
/* <integer> values */
reading-order: 1;
reading-order: -1;

/* Global values */
reading-order: inherit;
reading-order: initial;
reading-order: revert;
reading-order: revert-layer;
reading-order: unset;
```

### Value

> The `reading-order` property takes a single {{cssxref("integer")}} value specifying which ordinal group an element belongs to.

Reviewer: ```suggestion
- {{cssxref("&lt;integer&gt;")}}
  - : Represents the ordinal group an element belongs to.
```

## Description

> The `reading-order` property lets the author change where in the reading order an element is read out or tabbed to. It enables authors to override the reading order set by the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property on its parent.

Reviewer: ```suggestion
The `reading-order` property lets the author define where in the reading order an element is read out or tabbed to, optionally overriding the reading order set by the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property on its parent.
```

Author: updated

> The default `reading-order` value is `0` for elements inside a block, flex, or grid container. This puts them all in the same ordinal group. Sibling elements are ordered starting from the lowest numbered ordinal group to the highest. Therefore, to cause an element to be read out after its siblings, you could set a `reading-order` value or `1` or more on it. To cause an element to be read out before its siblings, you could set a `reading-order` value or `-1` or less on it.

Reviewer: ```suggestion
For reading and navigation, elements inside a block, flex, or grid container in a container are sorted by ascending `reading-order` value and then by their source code order. Items not given an explicit `reading-order` value are assigned the default value of `0`. This puts them all in the same ordinal group by default.

Sibling elements are ordered starting from the lowest numbered ordinal group to the highest. Therefore, to cause an element to be read out after its siblings, you could set a `reading-order` value or `1` or more on it. To cause an element to be read out before its siblings, you could set a `reading-order` value or `-1` or less on it.

The `reading-order` defines the reading and tabbing order. This property has no effect on visual order. To alter the visual order of items in a container, see the {{cssxref("order")}} property.
```
how does this interact with tabindex?

Author: updated.

As for `tabindex`, that is a good question — I've tried a bit of experimentation, but found the results a little bit confusing.

@rachelandrew, can you give us a summary of how `reading-flow`/`reading-order` interact with `tabindex`?

Reviewer 2: Reading flow containers become focus scope owners. The [PR here](https://github.com/whatwg/html/pull/10613) contains the HTML definition.

I believe [this explainer](https://github.com/explainers-by-googlers/css-reading-flow/blob/main/reading-flow-vs-tabindex.md) is also correct, but I'd treat that PR as the source of truth,

Author: OK, so from reviewing @estelle's tests, it looks like I am right. I'm not convinced we need to add another demo to demonstrate that `reading-flow`/`reading-order` overrides tabindex and it is ignored while they are in effect. We should just add some explanation to each page to mention this.

Does that work for you, @rachelandrew, or have I got this wrong?

## Formal definition

{{cssinfo}}

## Examples

### Grid row order

In this example there are six {{htmlelement("a")}} elements, laid out on a grid. One of them has a lower `reading-order` value set, so will be read out before the others, regardless of source or display order.

#### HTML

The six `<a>` elements are contained inside a wrapper {{htmlelement("div")}}.

```html
<div class="wrapper">
  <a href="#">Item 1</a>

>   <a href="#">Item 2</a>

Reviewer: ```suggestion
  <a class="bottom" href="#">Item 2</a>
```
we can also include a +1 example

Author: updated

  <a href="#">Item 3</a>
  <a class="top" href="#">Item 4</a>
  <a href="#">Item 5</a>
  <a href="#">Item 6</a>
</div>
```

#### CSS

> On the `<div>`, we set the [grid-auto-flow](/en-US/docs/Web/CSS/grid-auto-flow) property to `dense`, therefore items may display out of source order. The `reading-order` property on the `<a>` element with a class of [top](/en-US/docs/Web/CSS/top) is set to `-1`, therefore "Item 4" will be the first item in reading flow. The remaining items will be visited in the order they display in rows, as the `<div>` element's {{cssxref("reading-flow")}} property is set to [grid-rows](/en-US/docs/Web/CSS/grid-row).

Reviewer: ```suggestion
On the `<div>`, we set the [grid-auto-flow](/en-US/docs/Web/CSS/grid-auto-flow) property to `dense`, therefore items may display out of source order. The `reading-order` property on the `<a>` element with a class of `top` is set to `-1` and `bottom` is set to `1`, therefore "Item 4" will be the first item in reading flow and "Item 2" will be the last. The remaining items will be visited in the order they display in rows, as the `<div>` element's {{cssxref("reading-flow")}} property is set to [grid-rows](/en-US/docs/Web/CSS/grid-row).
```

Author: changed

```css
.wrapper {
  display: grid;
  grid-template-columns: repeat(3, 150px);
  grid-auto-flow: dense;
  reading-flow: grid-rows;
}

.top {
  reading-order: -1;

> }

Reviewer: add .bottom

Author: added

```

#### Result

The above demo renders as follows:

{{EmbedLiveSample('Grid row order', '100%', '100px')}}

> Try tabbing through the links, and note how "Item 4" is tabbed to first because of its modified `reading-order` value. After that, the items are tabbed to in grid row order.

Reviewer: ```suggestion
Try tabbing through the links, and note how "Item 4" is tabbed to first because of its modified `reading-order` value, while "number 2" is tabbed to last. Otherwise, the items are tabbed to in grid row order.
```

Author: changed

### Source order override

In this example there are five {{htmlelement("a")}} elements, displayed as block-level elements. One of them has a lower `reading-order` value set, so will be read out before the others, regardless of source or display order.

#### HTML

The five `<a>` elements are contained inside a wrapper {{htmlelement("div")}}.

```html
<div class="wrapper">
  <a href="#">Item 1</a>
  <a href="#">Item 2</a>
  <a href="#">Item 3</a>
  <a href="#">Item 4</a>
  <a href="#">Item 5</a>
</div>
```

#### CSS

The `<div>` element's `reading-flow` property is set to `source-order`, which allows `reading-order` to be used to override the default source reading order. The third `<a>` element has a `reading-order` value of `-1`, therefore it is read out first, regardless of source order.

```css
.wrapper {
  reading-flow: source-order;
}

.wrapper > a {
  display: block;
}

.wrapper a:nth-child(3) {

>   reading-order: -1;

Reviewer: consider adding a more negative value, demonstrating you can use numbers other than 1, and you can reorder more than one value in the same direction. Alternatively, give two links the same -1 vale, then point out how betwen x and 3, with the same reading order, they're read in source-order

Author: I added a more positive value for `bottom` instead.

---
title: reading-order
slug: Web/CSS/reading-order
page-type: css-property
browser-compat: css.properties.reading-order
---

{{CSSRef}}

> The **`reading-order`** [CSS](/en-US/docs/Web/CSS) property enables defining a {{glossary("reading order")}} for elements within a [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) container, which overrides any [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) set on the container via the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property.

Author: updated

## Syntax

```css
/* <integer> values */
reading-order: 1;
reading-order: -1;

/* Global values */
reading-order: inherit;
reading-order: initial;
reading-order: revert;
reading-order: revert-layer;
reading-order: unset;
```

### Value

- {{cssxref("&lt;integer&gt;")}}
  - : Represents the ordinal group an element belongs to.

## Description

> The `reading-order` property lets the author define where in the reading order an element is read out or tabbed to, optionally overriding the reading order set by the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property on its parent.

Reviewer: ```suggestion
The `reading-order` property lets the author define where in the reading order an element is read out or tabbed to,  relative to its siblings, within a reading flow container parent, optionally overriding the reading order set by the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property on that parent.
```

Author: updated

> For reading and navigation, elements inside a block, flex, or grid container in a container are sorted by ascending `reading-order` value and then by their source code order. Items not given an explicit `reading-order` value are assigned the default value of `0`. This puts them all in the same ordinal group by default.

Reviewer: ```suggestion
For reading and navigation, elements inside a block, flex, or grid container are sorted by ascending `reading-order` value. If multiple sibling have the same `reading-order` value, those items are sorted according to the containers `reading-flow` value. Sibling items not given an explicit `reading-order` value are assigned the default value of `0`, which puts all the sibling children of a reading flow container parent in the same ordinal group by default.
```

Author: updated

Sibling elements are ordered starting from the lowest numbered ordinal group to the highest. Therefore, to cause an element to be read out after its siblings, you could set a `reading-order` value or `1` or more on it. To cause an element to be read out before its siblings, you could set a `reading-order` value or `-1` or less on it.

> The `reading-order` defines the reading and tabbing order. This property has no effect on visual order. To alter the visual order of items in a container, use the {{cssxref("order")}} property.

Reviewer: ```suggestion
The `reading-order` defines the reading and tabbing order, but has no effect on visual order.
```
removed " To alter the visual order of items in a container, use the {{cssxref("order")}} property." because we don't want people to use order, but, if we do use order, we should likely list other ways the visual order is changes, like positioning, floating, grid-placement, reversing columns and rows, etc.

Author: updated

```html
<div class="wrapper">
  <a href="#">Item 1</a>
  <a class="bottom" href="#">Item 2</a>
  <a href="#">Item 3</a>
  <a class="top" href="#">Item 4</a>
  <a href="#">Item 5</a>
  <a href="#">Item 6</a>
</div>
```

#### CSS

On the `<div>`, we set the [grid-auto-flow](/en-US/docs/Web/CSS/grid-auto-flow) property to `dense`, therefore items may display out of source order. The `reading-order` property on the `<a>` element with a class of `top` is set to `-1`, therefore "Item 4" will be the first item in reading flow. The `reading-order` property on the `<a>` element with a class of `bottom` is set to `21`, therefore "Item 4" will be the last item in the reading flow. The remaining items will be visited in between, in the order they display in rows, as the `<div>` element's {{cssxref("reading-flow")}} property is set to [grid-rows](/en-US/docs/Web/CSS/grid-row).

```css
.wrapper {
  display: grid;
  grid-template-columns: repeat(3, 150px);
  grid-auto-flow: dense;
  reading-flow: grid-rows;
}

.top {
  reading-order: -1;
}

.bottom {
  reading-order: 21;
}
```

#### Result

The above demo renders as follows:

{{EmbedLiveSample('Grid row order', '100%', '100px')}}

> Try tabbing through the links, and note how "Item 4" is tabbed to first and "Item 2" is tabbed to last, because of their modified `reading-order` values. In between, the items are tabbed to in grid row order.

Reviewer: ```suggestion
Try tabbing through the links. Note how "Item 4" is tabbed to first and "Item 2" is tabbed to last because of their modified `reading-order` values. In between, the items are tabbed to in grid row order.
```
optionally "`grid-rows` order"

Author: updated

### Source order override

In this example there are five {{htmlelement("a")}} elements, displayed as block-level elements. One of them has a lower `reading-order` value set, so will be read out before the others, regardless of source or display order.

#### HTML

The five `<a>` elements are contained inside a wrapper {{htmlelement("div")}}.

```html
<div class="wrapper">
  <a href="#">Item 1</a>
  <a href="#">Item 2</a>
  <a href="#">Item 3</a>
  <a href="#">Item 4</a>
  <a href="#">Item 5</a>
</div>
```

#### CSS

The `<div>` element's `reading-flow` property is set to `source-order`, which allows `reading-order` to be used to override the default source reading order. The third `<a>` element has a `reading-order` value of `-1`, therefore it is read out first, regardless of source order.

```css
.wrapper {
  reading-flow: source-order;
}

.wrapper > a {
  display: block;
}

> .wrapper a:nth-child(3) {

Reviewer: ```suggestion
.wrapper a:nth-child(odd) {
```
if we want to show "all the elements in the ordinal group come first, in source order"

Author: This is quite a nice little change. I've done this and modified the surrounding explanation to suit.

---
title: reading-order
slug: Web/CSS/reading-order
page-type: css-property
browser-compat: css.properties.reading-order
---

{{CSSRef}}

> The **`reading-order`** [CSS](/en-US/docs/Web/CSS) property enables defining a {{glossary("reading order")}} for the children of a [reading flow container](/en-US/docs/Glossary/Reading_order#reading_flow_container).

Author: updated

> When an element's [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) parent container has a {{cssxref("reading-flow")}} resolving to a value other than `normal`, the `reading-order` property can be used to set the element's [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) position relative to that of its siblings.

Author: I ended up removing this paragraph from here, and putting it in the description instead. It was basically repeating the same information as the opening paragraph, but using more detail.

> When an element's [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) parent container has a {{cssxref("reading-flow")}} resolving to a value other than `normal`, the `reading-order` property can be used to set the element's [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) position relative to that of its siblings.

Author: I just tried adding:

```css
a:last-child {
  reading-order: -1;
}
```

to the flex and grid examples in the `reading-flow` article, and it worked as expected - the last item was always read out first (provided a non-normal `reading-flow` value was chosen); after that, the order was respected based on the `reading-flow` value.

> When an element's [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) parent container has a {{cssxref("reading-flow")}} resolving to a value other than `normal`, the `reading-order` property can be used to set the element's [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) position relative to that of its siblings.

Reviewer: `<a>` is not a block level item. In flex and grid, it becomes a flex-item or grid-item, so should respect -1, but if the container is a `<p>`, based on what I wrote above, it should not. see the codepen I posted - containing a paragraph with text and links.

> When an element's [block](/en-US/docs/Glossary/Block/CSS), [flex](/en-US/docs/Web/CSS/CSS_flexible_box_layout), or [grid](/en-US/docs/Web/CSS/CSS_grid_layout) parent container has a {{cssxref("reading-flow")}} resolving to a value other than `normal`, the `reading-order` property can be used to set the element's [reading flow](/en-US/docs/Glossary/Reading_order#reading_flow) position relative to that of its siblings.

Author: Done: https://issues.chromium.org/issues/420339599

## Syntax

```css
/* <integer> values */
reading-order: 1;
reading-order: -1;

/* Global values */
reading-order: inherit;
reading-order: initial;
reading-order: revert;
reading-order: revert-layer;
reading-order: unset;
```

### Value

- {{cssxref("&lt;integer&gt;")}}

>   - : Represents the ordinal group an element belongs to.

Author: updated

## Description

> The `reading-order` property lets the author define where in the reading order an element is read out or tabbed to relative to its siblings within a reading flow container parent, optionally overriding the reading order set by the [reading-flow](/en-US/docs/Web/CSS/reading-flow) property on that parent.

Author: This is different now. I replaced it with the second paragraph from the intro.

> For reading and navigation, elements inside a block, flex, or grid container are sorted by ascending `reading-order` value. If multiple siblings have the same `reading-order` value, those items are sorted according to the container's `reading-flow` value. Siblings not given an explicit `reading-order` value are assigned the default value of `0`, which puts all the children of a reading flow container in the same ordinal group by default.

Author: I think "value" is better here. No change.

> For reading and navigation, elements inside a block, flex, or grid container are sorted by ascending `reading-order` value. If multiple siblings have the same `reading-order` value, those items are sorted according to the container's `reading-flow` value. Siblings not given an explicit `reading-order` value are assigned the default value of `0`, which puts all the children of a reading flow container in the same ordinal group by default.

Reviewer: the `reading-flow` value creates a reading order for its children. The items follow that reading order, not a value, unless the `reading-order` changes the order `

> For reading and navigation, elements inside a block, flex, or grid container are sorted by ascending `reading-order` value. If multiple siblings have the same `reading-order` value, those items are sorted according to the container's `reading-flow` value. Siblings not given an explicit `reading-order` value are assigned the default value of `0`, which puts all the children of a reading flow container in the same ordinal group by default.

Author: I see what you mean, but I think both of us are right, really. You are looking at it in terms of the effect the property value has, and I am looking at it in terms of the thing that is used to create that effect (the property value).

I just think "`reading-flow` order" sounds a bit wordy and awkward. I have decided to just remove the word "value" altogether. "...sorted according to the container's `reading-flow`." works well, IMO.

> For reading and navigation, elements inside a block, flex, or grid container are sorted by ascending `reading-order` value. If multiple siblings have the same `reading-order` value, those items are sorted according to the container's `reading-flow` value. Siblings not given an explicit `reading-order` value are assigned the default value of `0`, which puts all the children of a reading flow container in the same ordinal group by default.

Reviewer: agree

Sibling elements are ordered starting from the lowest numbered ordinal group to the highest. Therefore, to cause an element to be read out after its siblings, you could set a `reading-order` value or `1` or more on it. To cause an element to be read out before its siblings, you could set a `reading-order` value or `-1` or less on it.

> The `reading-order` defines the reading and tabbing order, but has no effect on visual order.

Reviewer: I think we may want to add something to the effect of:

"The are HTML attributes and CSS properties that can make the reading order of an element and its children diverge from the visual order of a that element and its children, including [tablindex], [order], [grid-template-areas], and [grid-area]. The `reading-flow` property enables resetting the reading order for the children of a container, while the `reading-order` enables tweaking the position of individual elements within that reading flow."

Maybe as an intro to the description of both properties, since we don't have a guide.

> The `reading-order` defines the reading and tabbing order, but has no effect on visual order.

Author: Again, let's wait and see what @rachelandrew says about this.

## Formal definition

{{cssinfo}}

## Examples

### Grid row order

> In this example there are six {{htmlelement("a")}} elements, laid out on a grid. One of them has a lower `reading-order` value set, so will be read out before the others, and another has a higher `reading-order` value set, so will be read out after the others, regardless of source or display order.

Reviewer: ```suggestion
This example demonstrates using `reading-order` to control the positions of individual grid items within a grid container's reading order. One grid item has a `reading-order` value that is lower than the default `0`, so will be read out before it's sibling elements. Another has a higher `reading-order` value set, so will be read out after the others, regardless of source or display order.
```
this section intros the example, rather than explaining the code.

Author: updated

#### HTML

> The six `<a>` elements are contained inside a wrapper {{htmlelement("div")}}.

Reviewer: ```suggestion
The six {{htmlelement("a")}}  elements are contained inside a wrapper {{htmlelement("div")}}.
```

Author: updated

```html
<div class="wrapper">
  <a href="#">Item 1</a>
  <a class="bottom" href="#">Item 2</a>
  <a href="#">Item 3</a>
  <a class="top" href="#">Item 4</a>
  <a href="#">Item 5</a>
  <a href="#">Item 6</a>
</div>
```

#### CSS

> On the `<div>`, we set the [grid-auto-flow](/en-US/docs/Web/CSS/grid-auto-flow) property to `dense`, therefore items may display out of source order. The `reading-order` property on the `<a>` element with a class of `top` is set to `-1`, therefore "Item 4" will be the first item in reading flow. The `reading-order` property on the `<a>` element with a class of `bottom` is set to `21`, therefore "Item 4" will be the last item in the reading flow. The remaining items will be visited in between, in the order they display in rows, as the `<div>` element's {{cssxref("reading-flow")}} property is set to [grid-rows](/en-US/docs/Web/CSS/grid-row).

Reviewer: ```suggestion
On the `<div>`, we set the [grid-auto-flow](/en-US/docs/Web/CSS/grid-auto-flow) property to `dense`, therefore items may display out of source order. The `reading-order` property on the `<a>` element with a class of `top` is set to `-1`, therefore "Item 4" will be the first item in reading flow. The `reading-order` property on the `<a>` element with a class of `bottom` is set to `21`, therefore "Item 4" will be the last item in the reading order. The remaining items will be visited in between, in grid row order, as the `<div>` element's {{cssxref("reading-flow")}} property is set to [grid-rows](/en-US/docs/Web/CSS/grid-row).
```

Author: updated

```css
.wrapper {
  display: grid;
  grid-template-columns: repeat(3, 150px);
  grid-auto-flow: dense;
  reading-flow: grid-rows;
}

.top {
  reading-order: -1;
}

.bottom {
  reading-order: 21;
}
```

#### Result

The above demo renders as follows:

{{EmbedLiveSample('Grid row order', '100%', '100px')}}

Try tabbing through the links. Note how "Item 4" is tabbed to first and "Item 2" is tabbed to last because of their modified `reading-order` values. In between, the items are tabbed to in grid row order.

### Source order override

> In this example there are five {{htmlelement("a")}} elements, displayed as block-level elements. The odd-number items have a lower `reading-order` value set, so will be read out before the others, regardless of source or display order.

Reviewer: ```suggestion
This example demonstrates the grouping effect of the `reading-flow` property. Multipled items have the same negative `reading-order` value set, so will be read out as a group, in `reading-flow` order, before the other sibling elements, regardless of source or display order.
```
again, intro-ing the demo at the start, then explaining the code as it's presented.

#### HTML

The five `<a>` elements are contained inside a wrapper {{htmlelement("div")}}.

```html
<div class="wrapper">
  <a href="#">Item 1</a>
  <a href="#">Item 2</a>

>   <a href="#">Item 3</a>

Reviewer: Do we want to add something like 
```
  <a href="#" tabindex="5">Item 3</a>
```
and explain that it's ignored here? or should we add another example?

Author: Again, let's see what @rachelandrew says. I think we should add another example to explain the effect on tabindex.

  <a href="#">Item 4</a>
  <a href="#">Item 5</a>
</div>
```

#### CSS

> The `<div>` element's `reading-flow` property is set to `source-order`, which allows `reading-order` to be used to override the default source reading order. The odd-numbered `<a>` elements have a `reading-order` value of `-1`, therefore they are read out before the even-numbered items, regardless of source order.

Reviewer: ```suggestion
The `<div>` element's `reading-flow` property is set to `source-order`, which allows `reading-order` to be used to override the default source reading order. The odd-numbered `<a>` elements have a `reading-order` value of `-1`, therefore they are read out before the even-numbered items.
```
duplicated above

Author: ok, removed.
